clients/src/main/java/org/apache/kafka/common/requests/OffsetFetchRequest.java

```diff
@@ -18,154 +18,93 @@
-    private static final Schema OFFSET_FETCH_REQUEST_V1 = OFFSET_FETCH_REQUEST_V0;
-
-    // V2 adds top-level error code to the response as well as allowing a null offset array to indicate fetch
-    // of all committed offsets for a group
-    private static final Field TOPICS_V2 = TOPICS.nullableWithFields(
-            TOPIC_NAME,
-            PARTITIONS_V0);
-    private static final Schema OFFSET_FETCH_REQUEST_V2 = new Schema(
-            GROUP_ID,
-            TOPICS_V2);
-
-    // V3 request is the same as v2. Throttle time has been added to v3 response
-    private static final Schema OFFSET_FETCH_REQUEST_V3 = OFFSET_FETCH_REQUEST_V2;
-
-    // V4 bump used to indicate that on quota violation brokers send out responses before throttling.
-    private static final Schema OFFSET_FETCH_REQUEST_V4 = OFFSET_FETCH_REQUEST_V3;
-
-    // V5 adds the leader epoch of the committed offset in the response
-    private static final Schema OFFSET_FETCH_REQUEST_V5 = OFFSET_FETCH_REQUEST_V4;
-
-    public static Schema[] schemaVersions() {
-        return new Schema[] {OFFSET_FETCH_REQUEST_V0, OFFSET_FETCH_REQUEST_V1, OFFSET_FETCH_REQUEST_V2,
-            OFFSET_FETCH_REQUEST_V3, OFFSET_FETCH_REQUEST_V4, OFFSET_FETCH_REQUEST_V5};
-    }
+
+    public final OffsetFetchRequestData data;
 
     public static class Builder extends AbstractRequest.Builder<OffsetFetchRequest> {
-        private static final List<TopicPartition> ALL_TOPIC_PARTITIONS = null;
-        private final String groupId;
-        private final List<TopicPartition> partitions;
+
+        public final OffsetFetchRequestData data;
 
         public Builder(String groupId, List<TopicPartition> partitions) {
             super(ApiKeys.OFFSET_FETCH);
-            this.groupId = groupId;
-            this.partitions = partitions;
+
+            Map<String, OffsetFetchRequestTopic> offsetFetchRequestTopicMap = new HashMap<>();
+            partitions.forEach(topicPartition -> {
+                String topicName = topicPartition.topic();
+                OffsetFetchRequestTopic topic = offsetFetchRequestTopicMap.getOrDefault(
+                    topicName, new OffsetFetchRequestTopic().setName(topicName));
+                topic.partitionIndexes().add(topicPartition.partition());
+                offsetFetchRequestTopicMap.put(topicName, topic);
+            });
+
+            this.data = new OffsetFetchRequestData()
+                            .setGroupId(groupId)
+                            .setTopics(new ArrayList<>(offsetFetchRequestTopicMap.values()));
         }
 
         public static Builder allTopicPartitions(String groupId) {
-            return new Builder(groupId, ALL_TOPIC_PARTITIONS);
+            return new Builder(groupId, Collections.emptyList());
         }
 
         public boolean isAllTopicPartitions() {
-            return this.partitions == ALL_TOPIC_PARTITIONS;
+            return this.data.topics().isEmpty();
```

Comment:
This is not right. The check should be against null. An empty list just means a request for no offsets. I think we need to have test cases for the handling of "all topic" requests.